File: OpenGSGLibrary/GameFilesParser/Scanner.vb
'    OpenSGSGLibrary is an open-source library for Grand Strategy Games
'    Copyright (C) 2019  Torben Kneesch
'
'    This program is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    This program is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Collections.Generic
Imports System.IO
Imports System.Text

Namespace Parser

    ''' <summary>
    ''' Prepares a text for parsing by converting it into a stream of tokens representing control characters,
    ''' keywords and values. Unnecessary characters are removed in the rpocess.
    ''' </summary>
    Public Class Scanner

        ''' <summary>
        ''' Main Scan method of the Scanner class.
        ''' </summary>
        ''' <param name="reader">Text in the form of an implementation of TextReader i.e. StringBuilder</param>
        ''' <returns></returns>
        Public Iterator Function Scan(reader As TextReader) As IEnumerator(Of Token)
            While reader.Peek() <> -1
                Dim nextChar As Char = Chr(reader.Peek())
                If Char.IsWhiteSpace(nextChar) Or Char.IsControl(nextChar) Then
                    reader.Read()
                ElseIf nextChar = "#" Then
                    ' "#" is comment, skip rest of line
                    reader.ReadLine()
                ElseIf Char.IsLetterOrDigit(nextChar) Then
                    Dim sval As String = ScanName(reader)
                    If IsNumeric(sval) Then
                        Yield Token.FromValue(Val(sval))
                    Else
                        Yield Token.FromString(sval)
                    End If
                ElseIf Char.IsSymbol(nextChar) Then
                    Yield ScanSymbol(reader)
                Else
                    Yield ScanOther(reader)
                End If
            End While
            Yield Token.FromKind(Kind.EOF)
        End Function

        Private Function ScanName(reader As TextReader) As String
            Dim sb = New StringBuilder()
            Dim nextCharCode = reader.Peek()
            While (Char.IsLetterOrDigit(Chr(nextCharCode)) Or Chr(nextCharCode) = "_")
                sb.Append(Chr(reader.Read()))
                nextCharCode = reader.Peek()
                If nextCharCode = -1 Then Return sb.ToString()
            End While
            Return sb.ToString()
        End Function

        Private Function ScanSymbol(reader As TextReader) As Token
            Dim currentChar As Char = Chr(reader.Read())
            Select Case currentChar
                Case "="
                    Return Token.FromKind(Kind.EQUAL)
                Case Else
                    Return Token.FromKind(Kind.UNKNOWN)
            End Select
        End Function

        Private Function ScanOther(reader As TextReader) As Token
            Dim currentChar As Char = Chr(reader.Read())
            Select Case currentChar
                Case "{"
                    Return Token.FromKind(Kind.LEFTBRACKET)
                Case "}"
                    Return Token.FromKind(Kind.RIGHTBRACKET)
                Case Else
                    Return Token.FromKind(Kind.UNKNOWN)
            End Select
        End Function

    End Class

End Namespace

File: OpenGSGLibrary/GameFilesParser/Scanner.cs
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace OpenGSGLibrary.GameFilesParser
{
    /// <summary>
    /// Prepares a text for parsing by converting it into a stream of tokens representing control characters,
    /// keywords and values. Unnecessary characters are removed in the process.
    /// </summary>
    public class Scanner
    {
        /// <summary>
        /// Main Scan method of the Scanner class.
        /// </summary>
        /// <param name="reader">Text in the form of an implementation of TextReader i.e. StringBuilder</param>
        /// <returns></returns>
        public IEnumerator<Token> Scan(TextReader reader)
        {
            while (reader.Peek() != -1)
            {
                var nextChar = (char)reader.Peek();
                if (char.IsWhiteSpace(nextChar) || char.IsControl(nextChar))
                {
                    reader.Read();
                }
                else if (nextChar == '#')
                {
                    reader.ReadLine();
                }
                else if (char.IsLetterOrDigit(nextChar))
                {
                    var sval = ScanName(reader);
                    if (IsSimpleNumber(sval))
                    {
                        // If it contains a decimal point, keep it as a string
                        if (sval.Contains('.'))
                        {
                            yield return Token.FromString(sval);
                        }
                        else
                        {
                            // Integer - parse and return as int
                            yield return Token.FromValue(
                                (int)
                                    double.Parse(
                                        sval,
                                        System.Globalization.CultureInfo.InvariantCulture
                                    )
                            );
                        }
                    }
                    else
                    {
                        yield return Token.FromString(sval);
                    }
                }
                else if (nextChar == '-') // <- Add this check for negative numbers
                {
                    // Check if it's a negative number
                    reader.Read(); // Consume the minus
                    if (reader.Peek() != -1 && char.IsDigit((char)reader.Peek()))
                    {
                        // It's a negative number
                        var sval = "-" + ScanName(reader);
                        if (IsSimpleNumber(sval))
                        {
                            // If it contains a decimal point, keep it as a string
                            if (sval.Contains('.'))
                            {
                                yield return Token.FromString(sval);
                            }
                            else
                            {
                                // Integer - parse and return as int
                                yield return Token.FromValue(
                                    (int)
                                        double.Parse(
                                            sval,
                                            System.Globalization.CultureInfo.InvariantCulture
                                        )
                                );
                            }
                        }
                        else
                        {
                            yield return Token.FromString(sval);
                        }
                    }
                    else
                    {
                        // It's just a minus sign
                        yield return Token.FromKind(Kind.UNKNOWN);
                    }
                }
                else if (nextChar == '"')
                {
                    yield return Token.FromString(ScanEmbeddedName(reader));
                }
                else if (char.IsSymbol(nextChar))
                {
                    yield return ScanSymbol(reader);
                }
                else
                {
                    yield return ScanOther(reader);
                }
            }
            yield return Token.FromKind(Kind.EOF);
        }

        /// <summary>
        /// Checks if a string represents a simple numeric value.
        /// Supports integers and simple floating-point numbers (e.g., 123, -5, 0.5, -0.5).
        /// Does NOT support dates (e.g., 1950.01.09) or scientific notation.
        /// </summary>
        /// <param name="value">String to check.</param>
        /// <returns>True if the string is a valid simple number.</returns>
        private bool IsSimpleNumber(string value)
        {
            if (string.IsNullOrEmpty(value))
                return false;

            int startIndex = 0;

            // Handle optional leading minus sign
            if (value[0] == '-')
            {
                if (value.Length == 1)
                    return false; // Just a minus is not a number
                startIndex = 1;
            }

            bool hasDecimalPoint = false;

            for (int i = startIndex; i < value.Length; i++)
            {
                char c = value[i];

                if (c == '.')
                {
                    // Already have a decimal point, or decimal point is at start/end
                    if (hasDecimalPoint || i == startIndex || i == value.Length - 1)
                        return false;
                    hasDecimalPoint = true;
                }
                else if (!char.IsDigit(c))
                {
                    return false;
                }
            }

            return true;
        }

        private string ScanName(TextReader reader)
        {
            var sb = new StringBuilder();
            var nextCharCode = reader.Peek();
            while (
                char.IsLetterOrDigit((char)nextCharCode)
                || nextCharCode == '_'
                || nextCharCode == '.'
            )
            {
                sb.Append((char)reader.Read());
                nextCharCode = reader.Peek();
                if (nextCharCode == -1)
                    return sb.ToString();
            }
            return sb.ToString();
        }

        private string ScanEmbeddedName(TextReader reader)
        {
            var sb = new StringBuilder();
            var nextCharCode = reader.Read();
            nextCharCode = reader.Read();
            while ((char)nextCharCode != '"')
            {
                sb.Append((char)nextCharCode);
                nextCharCode = reader.Read();
            }
            return sb.ToString();
        }

        private Token ScanSymbol(TextReader reader)
        {
            var currentChar = (char)reader.Read();
            switch (currentChar)
            {
                case '=':
                    return Token.FromKind(Kind.EQUAL);
                case '>':
                    if (reader.Peek() == '=')
                    {
                        _ = (char)reader.Read();
                        return Token.FromKind(Kind.GREATER_EQUAL);
                    }
                    else
                    {
                        return Token.FromKind(Kind.GREATER);
                    }
                case '<':
                    if (reader.Peek() == '=')
                    {
                        _ = (char)reader.Read();
                        return Token.FromKind(Kind.LESS_EQUAL);
                    }
                    else
                    {
                        return Token.FromKind(Kind.LESS);
                    }
                default:
                    return Token.FromKind(Kind.UNKNOWN);
            }
        }

        private Token ScanOther(TextReader reader)
        {
            var currentChar = (char)reader.Read();
            switch (currentChar)
            {
                case '{':
                    return Token.FromKind(Kind.LEFTBRACKET);
                case '}':
                    return Token.FromKind(Kind.RIGHTBRACKET);
                default:
                    return Token.FromKind(Kind.UNKNOWN);
            }
        }
    }
}
